x-pack/plugin/mapper-counted-keyword/src/main/java/org/elasticsearch/xpack/countedkeyword/CountedKeywordFieldMapper.java

```diff
@@ -0,0 +1,391 @@
+        ft.setDocValuesType(DocValuesType.SORTED_SET);
+        ft.setTokenized(false);
+        ft.setOmitNorms(true);
+        ft.setIndexOptions(IndexOptions.DOCS);
+        ft.freeze();
+        FIELD_TYPE = freezeAndDeduplicateFieldType(ft);
+    }
+
+    private static class CountedKeywordFieldType extends StringFieldType {
+
+        private final MappedFieldType countFieldType;
+
+        CountedKeywordFieldType(
+            String name,
+            boolean isIndexed,
+            boolean isStored,
+            boolean hasDocValues,
+            TextSearchInfo textSearchInfo,
+            Map<String, String> meta,
+            MappedFieldType countFieldType
+        ) {
+            super(name, isIndexed, isStored, hasDocValues, textSearchInfo, meta);
+            this.countFieldType = countFieldType;
+        }
+
+        @Override
+        public ValueFetcher valueFetcher(SearchExecutionContext context, String format) {
+            return SourceValueFetcher.identity(name(), context, format);
+        }
+
+        @Override
+        public IndexFieldData.Builder fielddataBuilder(FieldDataContext fieldDataContext) {
+            failIfNoDocValues();
+
+            return (cache, breakerService) -> new AbstractIndexOrdinalsFieldData(
+                name(),
+                CoreValuesSourceType.KEYWORD,
+                cache,
+                breakerService,
+                (dv, n) -> new KeywordDocValuesField(FieldData.toString(dv), n)
+            ) {
+
+                @Override
+                public LeafOrdinalsFieldData load(LeafReaderContext context) {
+                    final SortedSetDocValues dvValues;
+                    final BinaryDocValues dvCounts;
+                    try {
+                        dvValues = DocValues.getSortedSet(context.reader(), getFieldName());
+                        dvCounts = DocValues.getBinary(context.reader(), countFieldType.name());
+                    } catch (IOException e) {
+                        throw new UncheckedIOException("Unable to load " + CONTENT_TYPE + " doc values", e);
+                    }
+
+                    return new AbstractLeafOrdinalsFieldData(toScriptFieldFactory) {
+
+                        @Override
+                        public SortedSetDocValues getOrdinalsValues() {
+                            return new CountedKeywordSortedBinaryDocValues(dvValues, dvCounts);
+                        }
+
+                        @Override
+                        public long ramBytesUsed() {
+                            return 0; // Unknown
+                        }
+
+                        @Override
+                        public void close() {
+                            // nothing to close
+                        }
+                    };
+                }
+
+                @Override
+                public LeafOrdinalsFieldData loadDirect(LeafReaderContext context) {
+                    return loadDirect(context);
```

Comment:
I think this should invoke `load()`. Now this is recursive call.

Reply:
Oh, good catch. I'll fix this.

Reply:
This still is recursive?

Reply:
Ah, sorry! I've changed this already when I temporarily turned the anonymous class into an inner class but forgot to apply the change again when I undid that change. :facepalm: Corrected in https://github.com/elastic/elasticsearch/pull/101826/commits/a074204328ca40f1d1d7660e8fdaebefc96a08f2